Remove frontend in production (#8)
meilisearch/specifications: text/0004-optional-frontend.md

- Title: Optional Meilisearch Front-end
- Start Date: 2020/11/16
- specification PR:
- Meilisearch Issue: #411

# Optional Meilisearch Front-end

## First section: Feature Description and Interaction

### Summary

For testing purposes, MeiliSearch is shipped with a frontend that is served on the server's root. This provides the user with a basic GUI to perform searches on his document collection. While this is useful in a development environment, this is not necessarily desirable in a production environment. This specification propose to make this frontend optional thanks to binary setting.

### Motivation

In most production cases, there is no need for this frontend, and exposing it is a bad practice. We want to provide an opt-out option for this feature.

### Prior Art and R&D

N.A

### Explanation

> On MeiliSearch startup, we would like to pass a flag `--disable-web-gui` that disables the web user interface.

Author: What do we do then? @curquiza @LegendreM

> On MeiliSearch startup, we would like to pass a flag `--disable-web-gui` that disables the web user interface.

Reviewer: I incline to the @LegendreM suggestion that automatically disables it when the production mode is set.  Since exposing the front-end is a bad practice in production, disable it automatically could help our users, otherwise all the MeiliSearch production should have both settings: `--env production` and `--disable-web-gui`. `--env production` should be enough.
Plus, ut would enforce users to use the production mode and so that we could know who is really using MeiliSearch in production 🙂

### Impact on documentation

> This new feature adds a new flag to the binary, that needs to be documented [here](https://docs.meilisearch.com/guides/advanced_guides/configuration.html).

Reviewer 2: It should also be explained:
- In the [environment documentation](https://docs.meilisearch.com/guides/advanced_guides/configuration.html#environment) 
- In the [web interface documentation](https://docs.meilisearch.com/guides/advanced_guides/web_interface.html#web-interface)

- Title: Optional Meilisearch Front-end
- Start Date: 2020/11/16
- specification PR: #8
- Meilisearch Issue: #411

# Optional Meilisearch Front-end

## First section: Feature Description and Interaction

### Summary

For testing purposes, MeiliSearch is shipped with a frontend that is served at the server's root. This provides the user with a basic GUI to perform searches on his documents. While this is useful in a development environment, this is not necessarily desirable in a production environment. This specification proposes to remove this frontend in a development environment.

### Motivation

> This frontend was developed for testing purposes, and it should not be present in a production environment.

Reviewer 3: ```suggestion
This frontend was developed for testing purposes, and it should be disabled by default in a production environment.
```

### Prior Art and R&D

N.A

### Explanation

> When the option `--env` or `MEILI_ENV` is set to `production`, the frontend is disabled.
> 

Reviewer 3: ```suggestion
When the option `--env` or `MEILI_ENV` is set to `production`, the frontend is disabled unless the option `--enable-gui` or `MEILI_ENABLE_GUI` is set to `true`.
```

Author: Can't we wait until someone complains until we introduce a new option? I like @LegendreM because it doesn't introduce any change to the interface, and t probably fits in 99,9% of the usecases, wdyt @eskombro, @curquiza ?

Reviewer: I'm not against Sam's suggestion, but on the other hand, I also think if a user wants the interface in production, this is not production, this is for dev purposes.
So, I'm more on Marin's side: wait for complaints and specific use-cases before adding this option. But this is just my opinion.

Author: No matter what, adding an option does not change anything to what is done here, so if we all agree on this, we can merge this spec. If adding an option is needed, it can be done in another RFC, without conflicts with this one.

Reviewer 3: > Can't we wait until someone complains

Ok, I am mainly saying this thinking on Sandbox users first (always transposing that to any other possible automation tools, like DO Image, Saas, and this kind of things) where the USER is not the OWNER. 

- I can set a production environment to give production-ready instances to someone else, it doesn't suit me to not have the dashboard as it is the only visible+interactive part of the running MeiliSerch instance in the browser (doesn't require a terminal or any other normal tool).
- I do a demo, I want to set it In production, I can still use that web GUI. 
- I do a tutorial, I want it activated to have a GUI.

Why are you taking this from me? 😢 

[ok @MarinPostma , someone has complained now haha ;) ]

Reviewer 3: I am only asking for this if there is no implementation complexity

Author: What's the difference between prod and dev env again?

Reviewer 3: Actually checking the [docs](https://docs.meilisearch.com/guides/advanced_guides/configuration.html#environment) it is not quite much of a difference. But I don't know what will be built on top of that.

Author: @eskombro Is it ok for you to switch to a dev env, until we figure out what we want to do with this settings?

Reviewer 3: Yes, let's not block this, go ahead :)

### Impact on documentation

Documentation needs to be updated in:

- The [environment documentation](https://docs.meilisearch.com/guides/advanced_guides/configuration.html#environment)
- The [web interface documentation](https://docs.meilisearch.com/guides/advanced_guides/web_interface.html#web-interface)

## Second Section: Technical Specifications

### Architecture

N.A

### Implementation Details

> Pass an argument to `create_app` to conditionally enable the frontend service.

Reviewer: Is this still needed? Or is this referred to the @eskombro comment (https://github.com/meilisearch/specifications/pull/8#issuecomment-729676988)

Author: This is just implementation detail, it only concerns the core team 😉

Author: So not related

Reviewer: My baaaaad 😇
